File: example-item-list/src/androidTest/java/com/gurunars/item_list/example/ActivityMainTest.java
package com.gurunars.item_list.example;

import android.support.test.filters.LargeTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.view.View;
import android.view.ViewGroup;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.InstrumentationRegistry.getInstrumentation;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.Espresso.openActionBarOverflowOrOptionsMenu;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.hasDescendant;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class ActivityMainTest {

    public static Matcher<View> nthChildOf(final Matcher<View> parentMatcher, final int childPosition) {
        return new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("position " + childPosition + " of parent ");
                parentMatcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                if (!(view.getParent() instanceof ViewGroup)) return false;
                ViewGroup parent = (ViewGroup) view.getParent();

                return parentMatcher.matches(parent)
                        && parent.getChildCount() > childPosition
                        && parent.getChildAt(childPosition).equals(view);
            }
        };
    }

    @Rule
    public ActivityTestRule<ActivityMain> mActivityRule = new ActivityTestRule<>(
            ActivityMain.class);

    private void clickMenu(String text) {
        openActionBarOverflowOrOptionsMenu(getInstrumentation().getTargetContext());
        onView(withText(text)).perform(click());
    }

    @Test
    public void clickingClear_shouldShowEmptyListView() {
        clickMenu("Clear");
        onView(withText("Empty")).check(matches(isDisplayed()));
    }

    private void assertList(String...expectedItems) {
        for (int i=0; i < expectedItems.length; i++) {
            nthChildOf(withId(R.id.itemList), i).matches(
                    hasDescendant(withText(expectedItems[i])));
        }
    }

    @Test
    public void deletingItems_shouldLeadToPartialRemoval() {
        clickMenu("Delete items");
        assertList(
            "0 @ 0 [tiger]",
            "2 @ 0 [monkey]"
        );
    }

    @Test
    public void createItems_shouldAppendItemsToTheEnd() {
        clickMenu("Create items");
        assertList(
            "0 @ 0 [tiger]",
            "1 @ 0 [wolf]",
            "2 @ 0 [monkey]",
            "3 @ 0 [lion]",
            "4 @ 0 [tiger]",
            "5 @ 0 [wolf]",
            "6 @ 0 [monkey]",
            "7 @ 0 [lion]"
        );
    }

    @Test
    public void updateItems_shouldChangeSomeOfItems() {
        clickMenu("Update items");
        assertList(
            "0 @ 0 [tiger]",
            "1 @ 1 [wolf]",
            "2 @ 0 [monkey]",
            "3 @ 1 [lion]"
        );
    }

    @Test
    public void moveUp_shouldPutItemFromBottomToTop() {
        clickMenu("Move up");
        assertList(
            "3 @ 0 [lion]",
            "0 @ 0 [tiger]",
            "1 @ 0 [wolf]",
            "2 @ 0 [monkey]"
        );
    }

    @Test
    public void moveDown_shouldPutItemFromTopToBottom() {
        clickMenu("Move down");
        assertList(
            "1 @ 0 [wolf]",
            "2 @ 0 [monkey]",
            "3 @ 0 [lion]",
            "0 @ 0 [tiger]"
        );
    }

    @Test
    public void resetItems_shouldSetItemsToInitialList() {
        clickMenu("Reset items");
        assertList(
            "0 @ 0 [tiger]",
            "1 @ 0 [wolf]",
            "2 @ 0 [monkey]",
            "3 @ 0 [lion]"
        );
    }

    @Before
    public void before() {
        clickMenu("Reset items");
    }

    @After
    public void after() {
        clickMenu("Reset items");
    }

}

File: examples/item-list/src/test/kotlin/com/gurunars/item_list/example/ActivityMainTest.kt
package com.gurunars.item_list.example

import android.app.Activity
import android.support.v7.widget.RecyclerView
import android.widget.TextView
import com.gurunars.android_utils.Animator
import junit.framework.Assert.assertEquals
import junit.framework.Assert.assertNotNull
import org.jetbrains.anko.find
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.fakes.RoboMenuItem

@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class)
class ActivityMainRoboTest {

    private lateinit var activity: Activity

    @Before
    fun setUp() {
        Animator.enabled = false
        activity = Robolectric.setupActivity(ActivityMain::class.java)
        clickMenu(R.id.reset)
    }

    private fun clickMenu(id: Int) {
        activity.onOptionsItemSelected(RoboMenuItem(id))
    }

    @Test
    fun clickingClear_shouldShowEmptyListView() {
        clickMenu(R.id.clear)
        assertNotNull(activity.find(R.id.noItemsLabel))
    }

    private fun assertList(vararg expectedItems: String) {
        val recycler = activity.find<RecyclerView>(R.id.recyclerView)
        for (i in expectedItems.indices) {
            assertEquals(expectedItems[i], (recycler.getChildAt(i) as TextView).text.toString())
        }
    }

    @Test
    fun deletingItems_shouldLeadToPartialRemoval() {
        clickMenu(R.id.delete)
        assertList(
            "#0{TIGER @ 0}",
            "#2{MONKEY @ 0}"
        )
    }

    @Test
    fun createItems_shouldAppendItemsToTheEnd() {
        clickMenu(R.id.create)
        assertList(
            "#0{TIGER @ 0}",
            "#1{WOLF @ 0}",
            "#2{MONKEY @ 0}",
            "#3{LION @ 0}",
            "#4{TIGER @ 0}",
            "#5{WOLF @ 0}",
            "#6{MONKEY @ 0}",
            "#7{LION @ 0}"
        )
    }

    @Test
    fun updateItems_shouldChangeSomeOfItems() {
        clickMenu(R.id.update)
        assertList(
            "#0{TIGER @ 0}",
            "#1{WOLF @ 1}",
            "#2{MONKEY @ 0}",
            "#3{LION @ 1}"
        )
    }

    @Test
    fun moveUp_shouldPutItemFromBottomToTop() {
        clickMenu(R.id.moveUp)
        assertList(
            "#3{LION @ 0}",
            "#0{TIGER @ 0}",
            "#1{WOLF @ 0}",
            "#2{MONKEY @ 0}"
        )
    }

    @Test
    fun moveDown_shouldPutItemFromTopToBottom() {
        clickMenu(R.id.moveDown)
        assertList(
            "#1{WOLF @ 0}",
            "#2{MONKEY @ 0}",
            "#3{LION @ 0}",
            "#0{TIGER @ 0}"
        )
    }

    @Test
    fun resetItems_shouldSetItemsToInitialList() {
        clickMenu(R.id.reset)
        assertList(
            "#0{TIGER @ 0}",
            "#1{WOLF @ 0}",
            "#2{MONKEY @ 0}",
            "#3{LION @ 0}"
        )
    }


}